perception/lidar_centerpoint_tvm/lib/centerpoint_tvm.cpp

```diff
@@ -0,0 +1,270 @@
+// limitations under the License.
+
+#include "lidar_centerpoint_tvm/centerpoint_tvm.hpp"
+
+#include <centerpoint_backbone/inference_engine_tvm_config.hpp>
+#include <centerpoint_encoder/inference_engine_tvm_config.hpp>
+#include <lidar_centerpoint_tvm/centerpoint_config.hpp>
+#include <lidar_centerpoint_tvm/network/scatter.hpp>
+#include <lidar_centerpoint_tvm/preprocess/generate_features.hpp>
+#include <tier4_autoware_utils/math/constants.hpp>
+#include <tvm_utility/pipeline.hpp>
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+// configs' path may need to be reset
+auto config_en = model_zoo::perception::lidar_obstacle_detection::centerpoint_encoder::
+  onnx_centerpoint_encoder::config;
+auto config_bk = model_zoo::perception::lidar_obstacle_detection::centerpoint_backbone::
+  onnx_centerpoint_backbone::config;
+
+namespace autoware
+{
+namespace perception
+{
+namespace lidar_centerpoint_tvm
+{
+
+VoxelEncoderPreProcessor::VoxelEncoderPreProcessor(
+  const tvm_utility::pipeline::InferenceEngineTVMConfig & config,
+  const CenterPointConfig & config_mod)
+: max_voxel_size(config.network_inputs[0].second[0]),
+  max_point_in_voxel_size(config.network_inputs[0].second[1]),
+  encoder_in_feature_size(config.network_inputs[0].second[2]),
+  datatype_bytes(config.tvm_dtype_bits / 8),
+  config_detail(config_mod)
+{
+  encoder_in_features.resize(max_voxel_size * max_point_in_voxel_size * encoder_in_feature_size);
+  // Allocate input variable
+  std::vector<int64_t> shape_x{1, max_voxel_size, max_point_in_voxel_size, encoder_in_feature_size};
+  TVMArrayContainer x{
+    shape_x,
+    config.tvm_dtype_code,
+    config.tvm_dtype_bits,
+    config.tvm_dtype_lanes,
+    config.tvm_device_type,
+    config.tvm_device_id};
+  output = x;
+}
+
+TVMArrayContainerVector VoxelEncoderPreProcessor::schedule(const MixedInputs & voxel_inputs)
+{
+  /// generate encoder_in_features from the voxels
+  generateFeatures(
+    voxel_inputs.features, voxel_inputs.num_points_per_voxel, voxel_inputs.coords,
+    voxel_inputs.num_voxels, config_detail, encoder_in_features);
+
+  TVMArrayCopyFromBytes(
+    output.getArray(), encoder_in_features.data(),
+    max_voxel_size * max_point_in_voxel_size * encoder_in_feature_size * datatype_bytes);
+
+  return {output};
+}
+
+VoxelEncoderPostProcessor::VoxelEncoderPostProcessor(
+  const tvm_utility::pipeline::InferenceEngineTVMConfig & config)
+: max_voxel_size(config.network_outputs[0].second[0]),
+  encoder_out_feature_size(config.network_outputs[0].second[2]),
+  datatype_bytes(config.tvm_dtype_bits / 8)
+{
+  pillar_features.resize(max_voxel_size * encoder_out_feature_size);
+}
+
+std::vector<float32_t> VoxelEncoderPostProcessor::schedule(const TVMArrayContainerVector & input)
+{
+  TVMArrayCopyToBytes(
+    input[0].getArray(), pillar_features.data(), pillar_features.size() * datatype_bytes);
+
+  return pillar_features;
```

Comment:
This brings another question to me. Should we stick to `pipleline` class? Since it introduces overhead, maybe we can get rid of `preprocess` and `postprocess`, I feel like havng `InferenceEngineTVM` is enough, just like `TensorRT` implementation of `centerpoint`.

Reply:
For now, I'll use `share_ptr` as suggested.